lib/SILGen/SILGenType.cpp

```diff
@@ -694,6 +695,29 @@ SILFunction *SILGenModule::emitProtocolWitness(
   auto requirementInfo =
       Types.getConstantInfo(TypeExpansionContext::minimal(), requirement);
 
+  auto shouldUseDistributedThunkWitness =
+      // always use a distributed thunk for distributed requirements:
+      requirement.isDistributedThunk() ||
+      // for non-distributed requirements, which are however async/throws,
+      // and have a proper witness (passed typechecking), we can still invoke
+      // them on the distributed actor; but must do so through the distributed
+      // thunk as the call "through an existential" we never statically know
+      // if the actor is local or not.
+      (requirement.hasDecl() && requirement.getFuncDecl() && requirement.hasAsync() &&
+       !requirement.getFuncDecl()->isDistributed() &&
+       witnessRef.hasDecl() && witnessRef.getFuncDecl() &&
+       witnessRef.getFuncDecl()->isDistributed());
+  // We only need to use thunks when we go cross-actor:
+  shouldUseDistributedThunkWitness = shouldUseDistributedThunkWitness &&
+                                     getActorIsolation(requirement.getDecl()) !=
```

Comment:
This check isn't really doing what it looks like it's doing, because the only ways in which the actor isolation will be equivalent between the requirement and the witness is if both are non-isolated or if they are in the same (non-generic) global actor. The latter case doesn't matter for distributed right now.